### PR TITLE
fix: use ide wrapper for zed binary

### LINKED ISSUE #6: Use SDK with zed with escape option off
I used this env `ZED_FLATPAK_NO_ESCAPE=1` and `FLATPAK_ENABLE_SDK_EXT=llvm15`, but the later was never recognized (the terminal doesn't see clang at all)

### PR DESCRIPTION
**Pull request by jerome-benoit**
Follow up to PR https://github.com/flathub/dev.zed.Zed/pull/8 
Closes #6 by allowing flatpak binary zed to use user defined env variables for flatpak sdks

@ShobuPrime, @tomaswarynyca: Supporting languages with a toolchain that might use the DSOs shipped in the flatpak like rust and probably others need ide-wrapper integration (ABI incompatilibity caused mainly by openssl DSOs). Please review and merge.

### CODE REVIEW

**Comment by tomaswarynyca** [comment]
Do you have evidence that the wrapper is being used?

**Comment by jerome-benoit** [comment]
> Do you have evidence that the wrapper is being used?

After installing the latest PR flatpak, doing `flatpak run dev.zed.Zed` must print: 

```
zed-wrapper: Setting up NPM packages
zed-wrapper: Setting up Cargo packages
zed-wrapper: Setting up Python packages
zed-wrapper: Setting up Ruby packages
```

But the default wrapper behavior is not satisfying, zed is ran as a foreground processus.

**Comment by VarLad** [comment]
Testing this. I get endless printing of `zed-wrapper: Environment is already set up` (Zed doesn't start)

**Comment by jerome-benoit** [comment]
> Testing this. I get endless printing of `zed-wrapper: Environment is already set up` (Zed doesn't start)

The latest build should fix it.

**Comment by Erven2016** [comment]
Thanks for your nice job, rust-analyzer is finally working now.


> Build [148184 successful](https://buildbot.flathub.org/#/builders/6/builds/148184) To test this build, install it from the testing repository:
> 
> ```
> flatpak install --user https://dl.flathub.org/build-repo/131269/dev.zed.Zed.flatpakref
> ```

I just installed this test build and `org.freedesktop.Sdk.Extension.rust-stable` with version of 23.08, and added an environment variable `FLATPAK_ENABLE_SDK_EXT=*` via `Flatseal`, then rust-analyzer worked normally. 


Before(Zed will report an error when opening an rust/cargo project):

Screenshot from 2024-09-21 23-51-48

After(seemingly cargo and rust-analyzer are working properly):

Screenshot from 2024-09-22 00-43-46

**Comment by jerome-benoit** [comment]
> > ```
> > flatpak install --user https://dl.flathub.org/build-repo/131269/dev.zed.Zed.flatpakref
> > ```
> 
> I just installed this test build and `org.freedesktop.Sdk.Extension.rust-stable` with version of 23.08, and added an environment variable `FLATPAK_ENABLE_SDK_EXT=*` via `Flatseal`, then rust-analyzer worked normally.
> 
> Before(Zed will report an error when opening an rust/cargo project)
> After(seemingly cargo and rust-analyzer are working properly)

You confirm that with the rust toolchain installed on the host OS, zed is also properly able handle a rust code base? 

> Suggestion: updating `EXEC=zed-wrapper` in `${FLATPAK_ID}.desktop`, or zed-wrapper is not working by clicking desktop icon.

And also a documentation dedicated to zed instead of the default one.

**Comment by Erven2016** [comment]
>
> You confirm that with the rust toolchain installed on the host OS, zed is also properly able to handle a rust code base?
>

image

Nothing happens. I can't find any rust toolchains from host in flatpak sandbox :crying_cat_face:

**Comment by jerome-benoit** [comment]
> Nothing happens. I can't find any rust toolchains from host in flatpak sandbox 😿

I fail to understand: you confirm `rust-analyzer` is the OS host PATH (`which  rust-analyzer`)? 

You have a zed plugin that looks for `rust-analyzer` to start rust LSP? 

I do see why it would be an issue: deno plugin is looking for deno in the OS host PATH and start its LSP ...

**Comment by jerome-benoit** [comment]
> > Nothing happens. I can't find any rust toolchains from host in flatpak sandbox 😿
> 
> I fail to understand: you confirm `rust-analyzer` is in the OS host PATH (`which rust-analyzer`)?
> 

After low-level investigation: for whatever reason, the rust zed plugin use openssl DSO when invoking `rust- analyzer`. The issue is showing up in the flatpak because of containerization and lack of openssl library DSO but is not flatpak packaging related. The next zed stable version should the runtime dep against openssl.

**Comment by jerome-benoit** [comment]
> > > Nothing happens. I can't find any rust toolchains from host in flatpak sandbox 😿
> > 
> > 
> > I fail to understand: you confirm `rust-analyzer` is in the OS host PATH (`which rust-analyzer`)?
> 
> After low-level investigation: for whatever reason, the rust zed plugin use openssl DSO when invoking `rust-analyzer`. The issue is showing up in the flatpak because of containerization and lack of openssl library DSO but is not flatpak packaging related. The next zed stable version should not be runtime dep against openssl.

**Review by tomaswarynyca — APPROVED** [review]
LGTM!

**Comment by jerome-benoit** [comment]
> LGTM!

I can merge it with one approval but I really would like to also have @ShobuPrime review ...

**Comment by jerome-benoit** [comment]
@tomaswarynyca: Ok to merge it? I think it has waited for too long now.

**Comment by teohhanhui** [comment]
Re: rust-analyzer

With the current version, I'm getting errors such as:

```
rust-analyzer: proc-macro crate build data is missing dylib path
```

And the language server logs reveals the problem:

```
error: linker `cc` not found
```

I assume it's trying to find `cc` on the host, which is not the behaviour I'd like, as I'm trying to keep the host system clean.

Does this PR change anything in that regard? Will `rust-analyzer` now be able to run inside the Flatpak instead?

Zed log reveals that it's using its own downloaded `rust-analyzer`:

```
2024-10-12T21:27:25.020755331+08:00 [INFO] starting language server process. binary path: "/home/teohhanhui/.var/app/dev.zed.Zed/data/zed/languages/rust-analyzer/rust-analyzer-2024-10-07", working directory: "/home/teohhanhui/Projects/teohhanhui/hexciv", args: []
```

So it's definitely weird that it then attempts to run that on the host... Or is there at least a way to force `rust-analyzer` to be run inside the Flatpak?

**Comment by Erven2016** [comment]
> So it's definitely weird that it then attempts to run that on the host... Or is there at least a way to force `rust-analyzer` to be run inside the Flatpak?

You can try this, but now you should install rust-stable SDK version of 24.08:

> I just installed this test build and `org.freedesktop.Sdk.Extension.rust-stable` with version of 23.08, and added an environment variable `FLATPAK_ENABLE_SDK_EXT=*` via `Flatseal`, then rust-analyzer worked normally.

I am using it and no any issue reported.

**Comment by Erven2016** [comment]
After installing rust-stable SDK, you can check the path if rust toolchains is installed in flakpak:

image

**Comment by jerome-benoit** [comment]
> > So it's definitely weird that it then attempts to run that on the host... Or is there at least a way to force `rust-analyzer` to be run inside the Flatpak?
> 
> You can try this. 

The README.md should be clear enough to do so. If not, please do a PR to amend it.

**Comment by teohhanhui** [comment]
@Erven2016:
> You can try this, but now you should install rust-stable SDK version of 24.08:

Thank you. It seems to work correctly for me.

> After installing rust-stable SDK, you can check the path if rust toolchains is installed in flakpak:

That would not work. `FLATPAK_ENABLE_SDK_EXT` is processed by the wrapper script, so inside `flatpak run --command=sh dev.zed.Zed` it would not give the expected result.

**Comment by jerome-benoit** [comment]
> 
> > The README.md should be clear enough to do so. If not, please do a PR to amend it.
> 
> Should we mention the Rust / rust-analyzer use case specifically? I think it'll help with discoverability.

The statement explaining the flatpak is by default not escaping the sandbox and that can cause issue if sandboxing is expected, such as using SDK extensions.

**Comment by teohhanhui** [comment]
> The statement explaining the flatpak is by default not escaping the sandbox and that can cause issue if sandboxing is expected, such as using SDK extensions.

Sorry, I'm confused by your statement. Could you rephrase please?

If I understand correctly, the Zed editor in this Flatpak performs sandbox escape by default, right?

Are you saying we should add an explanation that the SDK extensions are sandboxed, and using them together with Zed's sandbox escaping might therefore give surprising results? (Unless it's what users like myself want.)

**Comment by jerome-benoit** [comment]
> > The statement explaining the flatpak is by default not escaping the sandbox and that can cause issue if sandboxing is expected, such as using SDK extensions.
> 
> Sorry, I'm confused by your statement. Could you rephrase please?
> 

Please read the README.md. If after reading it; you do not understand what are the default settings of the this flatpak and how to change them, please do a PR to refine the phrasing.
